### PR TITLE
GH-20407: [Go] Array Builder for REE arrays

### LINKED ISSUE #20407: [Go] Array Builder for RLE Arrays
This is based on ARROW-17708

**Reporter**: [Matthew Topol](https://issues.apache.org/jira/browse/ARROW-17709) / @zeroshade
**Assignee**: [Matthew Topol](https://issues.apache.org/jira/browse/ARROW-17709) / @zeroshade
#### Related issues:
- [[Go] RLE Arrays To/From JSON](https://github.com/apache/arrow/issues/32948) (incorporates)
#### PRs and other links:
- [GitHub Pull Request #14114](https://github.com/apache/arrow/pull/14114)

<sub>**Note**: *This issue was originally created as [ARROW-17709](https://issues.apache.org/jira/browse/ARROW-17709). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

### PR DESCRIPTION
**Pull request by zeroshade**
CC @zagto 
* Closes: #20407

### CODE REVIEW

File: go/arrow/array/encoded.go [hunk 1]
```diff
@@ -156,3 +158,160 @@ func arrayRunEndEncodedApproxEqual(l, r *RunEndEncoded, opt equalOption) bool {
...
+
+func (b *RunEndEncodedBuilder) ValueBuilder() Builder { return b.values }
+func (b *RunEndEncodedBuilder) Append(n uint64) {
+	b.finishRun()
+	b.addLength(n)
+}
+func (b *RunEndEncodedBuilder) ContinueRun(n uint64) {
+	b.addLength(n)
+}
+func (b *RunEndEncodedBuilder) AppendNull() {
```
Comment (lidavidm) [inline go/arrow/array/encoded.go:239]: Would it be useful to have an AppendNulls as well?

Comment (lidavidm) [reply go/arrow/array/encoded.go:239]: I suppose you can always do that in terms of ValueBuilder and Append though

**Review by lidavidm — COMMENTED** [review]
LGTM.

File: go/arrow/array/encoded.go (hunk 1 above)
Comment (zeroshade) [reply go/arrow/array/encoded.go:239]: yea, if i'm gonna add `AppendNulls` then I should add it to *all* the builders and make it part of the interface and i don't really wanna do that in this PR haha.

**Pull request merged**